README.md
[![LinkedIn][linkedin-shield]][linkedin-url]
[![Apache License][license-shield]][license-url]
[![X][X-sheild]][X-url]

![Banner](./Assets/Banner.png)

<p align="center">
  <a href="#video-">View Video</a>
    ·
  <a href="https://github.com/Elio-Aliaj/Automated-Website-and-CMS/issues">Report Bug</a>
    ·
  <a href="https://github.com/Elio-Aliaj/Automated-Website-and-CMS/issues">Request Feature</a>
</p>

<details>
  <summary>Table of Contents 📋</summary>
  <ol>
    <li>
      <a href="#about-the-project-">About The Project 🌱</a>
      <ul>
        <li><a href="#key-features-">Key Features 🔑</a></li>    
        <li><a href="#technology-stack-">Technology Stack 🧑‍💻</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started-">Getting Started 🏁</a>
      <ul>
        <li><a href="#prerequisites-">Prerequisites 📦</a></li>
        <li><a href="#installation-%EF%B8%8F">Installation ⬇️</a></li>
        <li><a href="#running-aw--cms-">Running AW & CMS 🚀</a>
          <ul>
        <li><a href="#configuration-%EF%B8%8F">Configuration ⚙️</a>
        <li><a href="#running-%EF%B8%8F">Running 🏃‍♂️</a>
        <li><a href="#generated-website-files-%EF%B8%8F">Generated Website Files ♻️</a>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#usage-%EF%B8%8F">Usage 🛠️</a>
      <ul>
        <li><a href="#video-">View Video 🎬</a></li>
        <li><a href="#screenshot-">Screenshot 📸</a></li>
      </ul>
    </li>
    <li><a href="#license-">License 🪪</a></li>
    <li><a href="#contact-">Contact 📞</a></li>
  </ol>
</details>

## About The Project 🌱

![Form & Web](./Assets/Form%20&%20Web.png)

Automated-Website-and-CMS is a powerful tool designed to simplify website creation without the need for coding. This project empowers users to effortlessly generate personalized websites by filling out a simple form

### Key Features 🔑

- **Form-Based Website Creation**: Easily create websites by providing information through a straightforward form

- **Dynamic Customization**: Once the website is generated, users have three flexible options for customizing segments:

  1. **Regenerate Segments**: Update segments based on the information initially provided in the form

  2. **Interactive Fixes**: Users can choose to fix a segment by providing specific instructions. This information, combined with the segment code, is sent to ChatGPT 3.5 Turbo for intelligent adjustments

  3. **Instruction-Only**: Users can instruct ChatGPT 3.5 Turbo on changes without providing the existing segment code, allowing for complete freedom in designing the section

### Technology Stack 🧑‍💻

- [![HTML5][HTML5-sheild]][HTML5-url]
- [![CSS3][CSS3-sheild]][CSS3-url]
- [![SCSS][SCSS-sheild]][SCSS-url]
- [![JavaScript][JavaScript-sheild]][JavaScript-url]
- [![NodeJs][NodeJs-sheild]][NodeJs-url]
- [![Json-Server][Json-Server-sheild]][Json-Server-url]
- [![GPT3.5][GPT3.5-sheild]][GPT3.5-url]
- [![Dall-e][Dall-e-sheild]][Dall-e-url]

## Getting Started 🏁

### Prerequisites 📦

Before getting started, ensure you have the necessary prerequisites:

- **API Key from OpenAI:** Register for an account on OpenAI and obtain your API key. You'll need this key to leverage the power of ChatGPT 3.5 Turbo for advanced segment adjustments and personalized transformations. Visit [OpenAI](https://openai.com/blog/openai-api) to get your API key

- **Node.js:** Make sure you have Node.js installed. If not, you can download and install it from [Node.js](https://nodejs.org/en)

### Installation ⬇️

1. Clone the repository

   ```bash
   git clone https://github.com/Elio-Aliaj/Automated-Website-and-CMS.git
   ```

2. Install packages

   ```powershell
   cd .\Automated-Website-and-CMS\
   ```

   ```powershell
   npm install
   ```

### Running AW & CMS 🚀

#### Configuration ⚙️

1. In the main project folder (Automated-Website-and-CMS), locate the `.env.example` file
2. Rename it to `.env`
3. Open the `.env` file:
   - Replace `ID` with your actual OpenAI API Organization Id.
   - Replace `KEY` with your actual OpenAI API Key.

#### Running 🏃‍♂️

1.  Run Json Server

    ```powershell
    cd .\Automated-Website-and-CMS\Data\
    ```

    ```powershell
    npx json-server --watch db.json --port 8000
    ```

    <details>
    <summary>
    This will be shown in the console:
    </summary>
    <code>
        
        \{^_^}/ hi!

        Loading db.json
        Done

        Resources
        http://localhost:8000/formData
        http://localhost:8000/response

        Home
        http://localhost:8000

        Type s + enter at any time to create a snapshot of the database
        Watching...

    </code>
    </details>

####

2.  Run NodeJS

    ```powershell
    cd '.\Automated-Website-and-CMS\AI Caller\'
    ```

    ```powershell
    node .\AI_Caller.js
    ```

    <details>
    <summary>
    This will be shown in the console:
    </summary>
    <code>
        
        Server listening at http://localhost:3000

    </code>
    </details>

####

`Keep this console in view, because it will update you with any changes made to the website or inform for any errors`

3.  Open the Form

    - Navigate to the `.\Automated-Website-and-CMS\Form UI` directory.
    - Locate the `User_interface.html` file and open it in your preferred web browser.
    - The form UI will guide you through the process of filling out the necessary information to generate your initial website.

    <details>
    <summary>
    This Form will show up on a website:
    </summary>
    <img src="Assets/Form.jpeg">
    </details>

####

4.  Customize the Website

    - When hovering over every section with the mouse, a gear ⚙️ icon will appear. Clicking the gear icon presents three ways of modifying the website:

      1. **Generate Section Again:**

         - This option will regenerate the current section with the data from the initial form.

      2. **Fix the Current Section:**

         - Choose this option to fix the current segment by providing instructions on how you want the segment to change. The provided instructions, combined with the segment code, are sent to ChatGPT 3.5 Turbo. After processing, ChatGPT 3.5 Turbo makes the necessary changes and returns the updated code.

      3. **Generate by Input:**

         - Users can provide instructions to ChatGPT 3.5 Turbo for changes without providing the existing segment code. This allows for complete freedom in designing the section.

        <details>
        <summary>
        This will show up:
        </summary>
        <img src="Assets/Edit.jpeg">
        </details>

          ⚠️ Do not make more than one change at a time otherwise you will get an 🛑 Error 429 🛑

####

#### Generated Website Files ♻️

After completing the website generation process, you can find all the generated files and images in `.\Automated-Website-and-CMS\Gen_Websit` directory:

- Navigate to this directory to access and manage the files created during the website generation. You'll find the components, pages, and images that make up your personalized website.

- Feel free to explore and make further customizations if needed. If you have any questions or encounter issues, please refer to the documentation or reach out to the project's support channels.
   <details>
  <summary>
   To remove Gear Icon ⚙️:
   </summary>

  - In the end of the `Gen_Web.html` file you should delete this lines:

     <code>

    ```html
    <!-- Your can delete from HERE to... -->
    <script src="../Edit_Website/Edit.js"></script>
    <script type="module">
      const socket = new WebSocket("ws://localhost:3000");
      socket.onmessage = (event) => {
        if (event.data === "reload") {
          location.reload();
        }
      };
    </script>
    <!-- ...to HERE after you are done with the website creation-->
    ```

     </code>
    </details>

## Usage 🛠️

### Video 🎬

https://github.com/Elio-Aliaj/Automated-Website-and-CMS/assets/119447071/13fdc1fa-43c7-423a-9f14-4427462cd6c7

### Screenshot 📸

<details>
<summary>
Here are some examples:
</summary>
<img src="Assets/Exaple 1.png">
<img src="Assets/Edit.jpeg">
<img src="Assets/Example 2.jpeg">
<img src="Assets/Form & Web.png">
<img src="Assets/Form SC.png">
<img src="Assets/Web Created.jpeg">
<img src="Assets/Example 3.png">
<img src="Assets/Example 4.jpeg">

</details>

## License 🪪

[![Apache License][license-shield]][license-url]

©️ Copyright 2023 Elio Aliaj <elio.aliaj0@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

## Contact 📞

[![Gmail][Gmail-sheild]][Gmail-url]
[![LinkedIn][linkedin-shield]][linkedin-url]
[![X][X-sheild]][X-url]
[![GitHub][GitHub-sheild]][GitHub-url]

<!-- Links & Logos -->

[linkedin-shield]: https://img.shields.io/badge/LinkedIn-%230A66C2?style=for-the-badge&logo=linkedin
[linkedin-url]: https://www.linkedin.com/in/elio-aliaj-48084a255/
[license-shield]: https://img.shields.io/github/license/Elio-Aliaj/Automated-Website-and-CMS?style=for-the-badge&logo=Apache&logoColor=%23D22128&color=%23D22128
[license-url]: https://github.com/Elio-Aliaj/Automated-Website-and-CMS/blob/main/LICENSE.md
[X-sheild]: https://img.shields.io/badge/Elio%20Aliaj-%23000000?style=for-the-badge&logo=x&labelColor=%23000000
[X-url]: https://twitter.com/Elio_Aliaj
[HTML5-sheild]: https://img.shields.io/badge/HTML5-E34F26?style=for-the-badge&logo=html5&logoColor=%23fff
[HTML5-url]: https://en.wikipedia.org/wiki/HTML5
[CSS3-sheild]: https://img.shields.io/badge/CSS3-1572B6?style=for-the-badge&logo=CSS3
[CSS3-url]: https://css3.com/
[SCSS-sheild]: https://img.shields.io/badge/SCSS-CC6699?style=for-the-badge&logo=sass&logoColor=%23fff
[SCSS-url]: https://sass-lang.com/documentation/syntax/
[JavaScript-sheild]: https://img.shields.io/badge/Javascript-F7DF1E?style=for-the-badge&logo=javascript&logoColor=%23fff
[JavaScript-url]: https://www.javascript.com/
[NodeJs-sheild]: https://img.shields.io/badge/Node.js-339933?style=for-the-badge&logo=nodedotjs&logoColor=%23fff
[NodeJs-url]: https://nodejs.org/en
[GPT3.5-sheild]: https://img.shields.io/badge/Chat%20GPT%203.5%20Turbo-412991?style=for-the-badge&logo=openai&logoColor=%23fff&label=Open%20ai
[GPT3.5-url]: https://openai.com/chatgpt
[Dall-e-sheild]: https://img.shields.io/badge/Dall--e%202-412991?style=for-the-badge&logo=openai&logoColor=%23fff&label=Open%20AI
[Dall-e-url]: https://openai.com/dall-e-2
[Json-Server-sheild]: https://img.shields.io/badge/Json%20Server-%23339933?style=for-the-badge&logo=npm&logoColor=%23fff&label=Npm&labelColor=%23CB3837
[Json-Server-url]: https://www.npmjs.com/package/json-server
[Gmail-sheild]: https://img.shields.io/badge/Gmail-%23EA4335?style=for-the-badge&logo=gmail&labelColor=%23fff
[Gmail-url]: mailto:elioaliaj0@gmail.com
[GitHub-sheild]: https://img.shields.io/badge/GitHub-%23181717?style=for-the-badge&logo=github
[GitHub-url]: https://github.com/Elio-Aliaj


CHANGELOG.md
# Changelog

## [0.1.3](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.1.2...v0.1.3) (2023-12-10)


### Bug Fixes

* fix license badge ([45a037c](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/commit/45a037cdc1ba6e7b6ad51cadbdc7a7ec94a62c47))

## [0.1.2](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.1.1...v0.1.2) (2023-12-10)


### Maintenance Activities

* add warning for error 429 ([2230dc7](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/commit/2230dc7e844ed1b2d1d58698ebaa9ba85bed3cb9))

## [0.1.1](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.1.0...v0.1.1) (2023-12-10)


### Maintenance Activities

* some fixes for api configuration ([a5bf0e2](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/commit/a5bf0e2bf75592f9ae441ff98377e94aeee7cc75))

## [0.1.0](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.0.2...v0.1.0) (2023-12-10)


### Features

* add edit functionality for website segments ([3253665](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/commit/32536658037ae9873743afcfe611ed1e391fbf51))

## [0.0.2](https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.0.2...v0.0.2) (2023-10-02)

### Bug fixes

- the code is more readable
- add many comments to explain the code
- tweak AI prompt to be more precise

### To Do

- Make a logo for the website with AI
- Make sections possible to edit

**Full Changelog**: https://github.com/Elio-Aliaj/Automated-Website-and-CMS/compare/v0.0.2...v0.0.2


Form UI/form.js
//Color checker
const form = document.querySelector("form");
const colorInput = document.querySelector("#web_color1");
const colorHiddenInput = document.querySelector("#color-hidden");

colorInput.addEventListener("change", function () {
  colorHiddenInput.value = colorInput.value;
});

form.addEventListener("submit", function (event) {
  if (colorHiddenInput.value === "#000000") {
    event.preventDefault();
    alert("Please pick at last a color");
  }
  if (colorHiddenInput.value !== "#000000") {
    event.preventDefault();
    const data = new FormData(this);
    var formData = {};
    for (const pair of data.entries()) {
      formData[pair[0]] = pair[1];
    }

    fetch("http://localhost:3000/submit-form", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(formData),
    });
    setTimeout(() => {
      window.location.href = "../Gen_Website/Gen_Web.html";
    }, 500);
  }
});

//Drop Down Menu
let names = [
  "Social Media",
  "E-commerce",
  "News",
  "Blogging",
  "Entertainment",
  "Search Engine",
  "Educational",
  "Job Search",
  "Forum",
  "Community",
  "Government",
  "Business Directory",
  "Weather",
  "Sports",
  "Travel",
  "Food",
  "Music",
  "Video Sharing",
  "Photo Sharing",
  "Online Marketplace",
  "Review Aggregator",
  "Video Game",
  "Gaming Community",
  "Personal Portfolio",
  "Online Magazine",
  "Streaming Video",
  "Live Streaming",
  "Online Learning",
  "Online Forum",
  "Dating",
  "File Sharing",
  "Webmail",
  "Online Auction",
  "Real Estate",
  "Internet Radio",
  "Tech News",
  "E-book",
  "Humor",
  "Science",
  "Fashion",
  "Health",
  "Beauty",
  "Automotive",
  "Hobby",
  "Outdoor Recreation",
  "Parenting",
  "Finance",
  "Insurance",
  "Social Network for Professionals",
  "Online Business Card",
  "Graphic Design",
  "Web Design",
  "Freelance Marketplace",
  "Online Translation",
  "Online Collaboration",
  "Virtual Tour",
  "Gossip",
  "Celebrity News",
  "Personal Blog",
  "Educational Games",
  "Personal Finance",
  "Sports Betting",
  "Stock Trading",
  "Online Advertising",
  "Political News",
  "Charity",
  "Non-profit",
  "Horoscope",
  "Weather Forecast",
  "Sports News",
  "Fantasy Sports",
  "Political Blog",
  "Music Streaming",
  "Podcast",
  "Gardening",
  "Recipe",
  "Humor Blog",
  "Online Comics",
  "Webcomic",
  "Artistic Portfolio",
  "Digital Art Community",
  "Photography Portfolio",
  "Travel Blog",
  "Tech Blog",
  "Job Board",
  "Creative Writing",
  "Literary Magazine",
  "Online Resume",
  "Online Store",
  "Beauty Blog",
  "Fitness Blog",
  "Pet Care",
  "Pet Adoption",
  "Animal Rights",
  "Home Improvement",
  "DIY",
  "Architecture",
  "Wedding Planning",
  "Photography Services",
  "Interior Design",
  "Greeting Card",
  "Online Therapy",
  "Legal Advice",
  "Medical Advice",
  "Health Forum",
  "Online Support Group",
  "Online Meditation",
  "Religious Forum",
  "Online Book Club",
  "Music News",
  "Online Jewelry Store",
  "Handmade Crafts",
  "Luxury Goods",
  "Web Hosting",
  "Online Banking",
  "Online Tax Filing",
  "Online Surveys",
  "Online Tutoring",
  "Online Gaming Store",
  "Social Network for Gamers",
  "Online Auction for Charity",
  "Online Fundraising",
  "Online Auction for Art",
  "Online Auction for Collectibles",
  "Online Auction for Antiques",
  "Online Auction for Cars",
  "Online Auction for Real Estate",
  "Online Auction for Travel Packages",
  "Online Auction for Luxury Goods",
  "Online Auction for Electronics",
  "Online Auction for Sports Memorabilia",
  "Online Auction for Toys",
  "Online Auction for Clothing",
  "Online Auction for Handmade Crafts",
  "Online Auction for Jewelry",
];
//Sort names in ascending order
let sortedNames = names.sort();

//reference
let input = document.getElementById("web_type");

//Execute function on keyup
input.addEventListener("keyup", (e) => {
  //loop through above array
  //Initially remove all elements ( so if user erases a letter or adds new letter then clean previous outputs)
  removeElements();
  for (let i of sortedNames) {
    //convert input to lowercase and compare with each string

    if (
      i.toLowerCase().startsWith(input.value.toLowerCase()) &&
      input.value != ""
    ) {
      //create li element
      let listItem = document.createElement("li");
      //One common class name
      listItem.classList.add("list-items");
      listItem.style.cursor = "pointer";
      listItem.setAttribute("onclick", "displayNames('" + i + "')");
      //Display matched part in bold
      let word = "<b>" + i.substr(0, input.value.length) + "</b>";
      word += i.substr(input.value.length);
      //display the value in array
      listItem.innerHTML = word;
      document.querySelector(".list").appendChild(listItem);
    }
  }
});
function displayNames(value) {
  input.value = value;
  removeElements();
}
function removeElements() {
  //clear all the item
  let items = document.querySelectorAll(".list-items");
  items.forEach((item) => {
    item.remove();
  });
}

// Name of Website Genaerator
let names2 = [
  "Facebook",
  "Amazon",
  "WordPress",
  "CNN",
  "Google",
  "Gmail",
  "YouTube",
  "Udemy",
  "Indeed",
  "eBay",
  "Match",
  "Bluehost",
  "Booking.com",
  "ESPN",
  "Weather.com",
  "Zillow",
  "NASDAQ",
  "ASOS",
  "Steam",
  "Spotify",
  "DeviantArt",
  "Food.com",
  "WebMD",
  "Politico",
  "ScienceDirect",
  "TechCrunch",
  "Gardening Know How",
  "Sephora",
  "Car and Driver",
  "IMDb",
  "Forbes",
  "Charity Navigator",
  "Project Gutenberg",
  "Podbean",
  "BabyCenter",
  "LinkedIn",
  "Upwork",
  "Duolingo",
  "Eventbrite",
  "Coinbase",
  "Wix",
  "Fitbit",
  "Oculus",
  "eBid",
  "Dropbox",
  "Reddit",
  "Kickstarter",
  "Delicious",
  "GoFundMe",
  "Flickr",
  "Twitch",
  "Shutterfly",
  "Google Analytics",
  "Namecheap",
  "Monster",
  "PayPal",
  "Mega",
  "StumbleUpon",
  "Patreon",
  "Netflix",
  "Vistaprint",
  "Google Drive",
  "HubSpot",
  "SurveyMonkey",
  "Zillow Rental Manager",
  "WordPress.com",
  "Canva",
  "TurboTax",
  "BetterHelp",
  "Chase",
  "Squarespace",
  "Facebook Ads Manager",
  "DocuSign",
  "Mint",
  "Noom",
  "Etsy",
  "Skillshare",
  "Carbonite",
  "Box",
  "Hootsuite",
  "Zoho",
  "Evernote",
  "MyFitnessPal",
  "TD Bank",
  "Bing",
  "Google Ads",
  "DocuSign Signature",
  "Mental Health America",
  "TaxSlayer",
  "eBay for Charity",
  "Walmart Marketplace",
  "Square",
  "Amazon Web Services",
  "Zoom",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
  "www",
];
//Sort names in ascending order
let sortedNames2 = names2.sort();

//reference
let input2 = document.getElementById("web_name");

//Execute function on keyup
input2.addEventListener("keyup", (e) => {
  //loop through above array
  //Initially remove all elements ( so if user erases a letter or adds new letter then clean previous outputs)
  removeElements();
  for (let i of sortedNames2) {
    //convert input to lowercase and compare with each string

    if (
      i.toLowerCase().startsWith(input2.value.toLowerCase()) &&
      input2.value != ""
    ) {
      //create li element
      let listItem2 = document.createElement("li");
      //One common class name
      listItem2.classList.add("list-items");
      listItem2.style.cursor = "pointer";
      listItem2.setAttribute("onclick", "displayNames2('" + i + "')");
      //Display matched part in bold
      let word = "<b>" + i.substr(0, input2.value.length) + "</b>";
      word += i.substr(input2.value.length);
      //display the value in array
      listItem2.innerHTML = word;
      document.querySelector(".list2").appendChild(listItem2);
    }
  }
});
function displayNames2(value2) {
  input2.value = value2;
  removeElements();
}
function removeElements() {
  //clear all the item
  let items = document.querySelectorAll(".list-items");
  items.forEach((item) => {
    item.remove();
  });
}


Changer/Modifying_Web.js
import cheerio from "cheerio";
import { readFileSync, writeFileSync } from "fs";
import sass from "node-sass";
import chalk from "chalk";
import webPartCreator from "./Web_Part_Creator.js";

export default async function modifyFile(
  aiResponse,
  webPart,
  extractHtmlCss,
  formData
) {
  //Changing CSS file with extracted code to the class with the name based on the web part

  //1# Emptying the SCSS file so new data can be inserted into it
  extractHtmlCss.stats = "NOT OK";
  const scssFile = "../Gen_Website/Gen_Web.scss";
  let cssString = readFileSync(scssFile, "utf8");
  let startDelimiter = "//$" + webPart + "***";
  let endDelimiter = "//!" + webPart + "***";
  let replacement = "\n." + webPart + "_AXZCAASD{\n}\n";
  //Find limits
  let startIndex = cssString.indexOf(startDelimiter) + startDelimiter.length;
  let endIndex = cssString.indexOf(endDelimiter);
  //Replace string
  if (startIndex - startDelimiter.length !== -1 && endIndex !== -1) {
    let textBetweenDelimiters = cssString.substring(startIndex, endIndex);
    let cleanCssString = cssString.replace(textBetweenDelimiters, replacement);

    //2# Replacing the empty part of the SCSS file with the CSS extracted
    let regexPattern = new RegExp(`(\\.${webPart}_AXZCAASD\\s*{)([^}]*)(})`);
    let finalCssString = cleanCssString.replace(
      regexPattern,
      `$1\n  ${extractHtmlCss.css}\n$3`
    );
    writeFileSync("../Gen_Website/Gen_Web.scss", finalCssString);
  } else {
    console.log(chalk.red("Something is wrong with Gen_Web.scss file"));
  }
  //3# Compiling SCSS to CSS
  function renderSaa() {
    return new Promise((resolve, reject) => {
      sass.render(
        {
          file: "../Gen_Website/Gen_Web.scss",
          outFile: "../Gen_Website/Gen_Web.css",
        },
        async function (error, result) {
          if (!error) {
            writeFileSync("../Gen_Website/Gen_Web.css", result.css);
            resolve("Successful");
          } else {
            console.error(chalk.red(error), chalk.yellow(" Trying again..."));
            if (formData.reTry < 4) {
              formData.reTry++;
              resolve(await webPartCreator(formData, webPart));
            } else if (formData.reTry == 4) {
              console.log(chalk.red("Rendering code fail!"));
              reject("Failed");
            }
          }
        }
      );
    });
  }
  return renderSaa()
    .then(async (status) => {
      if (status == "Successful") {
        //Changing HTML file with extracted code to the class with the name based on the web part
        const html = readFileSync("../Gen_Website/Gen_Web.html", "utf8");
        let $ = cheerio.load(html);
        $("." + webPart + "_AXZCAASD").html(extractHtmlCss.html);
        writeFileSync("../Gen_Website/Gen_Web.html", $.html());

        await saveData(status, aiResponse, webPart);
      }
      return status;
    })
    .catch((error) => {
      return error;
    });
}

async function saveData(status, aiResponse, webPart) {
  if (status == "Successful") {
    var dataSave = { response: aiResponse.completion.content, id: webPart };
    await fetch(`http://localhost:8000/response/${webPart}`).then(
      async (res) => {
        if (res.ok) {
          await fetch(`http://localhost:8000/response/${webPart}`, {
            method: "PUT",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(dataSave),
          }).then(() => {
            console.log(
              chalk.hex(`#29D479`)(
                `Success, ${webPart} responds updated to Database`
              )
            );
          });
        } else {
          await fetch(`http://localhost:8000/response`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(dataSave),
          }).then(() => {
            console.log(
              chalk.hex(`#29D479`)(
                `Success, ${webPart} responds added to Database`
              )
            );
          });
        }
      }
    );
  }
}


Changer/Extract_Code_Respond.js
import chalk from "chalk";
import webPartCreator from "./Web_Part_Creator.js";

export default async function extractCode(aiResponse, formData, webPart) {
  const respondToText = aiResponse.completion.content;

  //Extracting the code from the response putting it in an object and returning this object

  const htmlStartTag = "$HTML";
  const htmlEndTag = "!HTML";
  var extractedHTML = "";
  const cssStartTag = "$CSS";
  const cssEndTag = "!CSS";
  var extractedCSS = "";
  var extractHtmlCss = {};

  const htmlStartIndex = respondToText.indexOf(htmlStartTag);
  const cssStartIndex = respondToText.indexOf(cssStartTag);
  if (htmlStartIndex !== -1 && cssStartIndex !== -1) {
    const htmlEndIndex = respondToText.indexOf(
      htmlEndTag,
      htmlStartIndex + htmlStartTag.length
    );
    const cssEndIndex = respondToText.indexOf(
      cssEndTag,
      cssStartIndex + cssStartTag.length
    );

    if (htmlEndIndex !== -1 && cssEndIndex !== -1) {
      extractedHTML = respondToText.substring(
        htmlStartIndex + htmlStartTag.length,
        htmlEndIndex
      );
      extractedCSS = respondToText.substring(
        cssStartIndex + cssStartTag.length,
        cssEndIndex
      );
      extractHtmlCss.html = extractedHTML;
      extractHtmlCss.css = extractedCSS;
      extractHtmlCss.stats = "OK";
    } else {
      extractHtmlCss.stats = "NOT OK";
    }
  } else {
    extractHtmlCss.stats = "NOT OK";
  }

  //Check up to 3 time if the the tags can be found in not error displayed

  if (extractHtmlCss.stats == "NOT OK" && formData.reTry < 5) {
    console.log(chalk.red("No tag found."), chalk.yellow("Trying again..."));
    formData.reTry++;
    await webPartCreator(formData, webPart);
  } else if (formData.reTry == 5) {
    console.log(chalk.red("Extracting code fail!"));
    extractHtmlCss.stats = "NOT OK";
  }
  return extractHtmlCss;
}


AI Caller/AI_Caller.js
import { Configuration, OpenAIApi } from "openai";
import express from "express";
import bodyParser from "body-parser";
import cors from "cors";
import webPartCreator from "../Changer/Web_Part_Creator.js";
import { config } from "dotenv";
config({ path: "../.env" });
import chalk from "chalk";
import http from "http";
import WebSocket from "ws";

const configuration = new Configuration({
  organization: process.env.ORGANIZATION, //get ID form OpenAI
  apiKey: process.env.API_KEY, //get the key form OpenAI
});

const openai = new OpenAIApi(configuration);

const app = express();
const server = http.createServer(app);
const wss = new WebSocket.Server({ server });
const port = 3000;

app.use(bodyParser.json());
app.use(cors());

app.use(bodyParser.urlencoded({ extended: true }));

//Getting data from Form (User_Interface.html) after submission and processing to create every part of the website
app.post("/submit-form", async (formData, res) => {
  formData = formData.body;
  if (formData.ReStatus == "Generate Again") {
    var webParts = [formData.section];
    formData.formData.ReStatus = formData.ReStatus;
    formData = formData.formData;
  } else if (formData.ReStatus == "Fix the code") {
    var webParts = [formData.section];
    formData.formData.ReStatus = formData.ReStatus;
    formData.formData.code = formData.code;
    formData.formData.instructions = formData.instructions;
    formData = formData.formData;
  } else if (formData.ReStatus == "User preferences") {
    var webParts = [formData.section];
    formData.formData.ReStatus = formData.ReStatus;
    formData.formData.instructions = formData.instructions;
    formData = formData.formData;
  } else {
    // Saving form data to database
    formData.id = 1;
    await fetch(`http://localhost:8000/FormData/${formData.id}`).then(
      async (res) => {
        if (res.ok) {
          await fetch(`http://localhost:8000/FormData/${formData.id}`, {
            method: "PUT",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(formData),
          }).then(() => {
            console.log(
              chalk.hex(`#137952`)(
                "Success, Submitted Form Data Updated to Database"
              )
            );
          });
        } else {
          await fetch(`http://localhost:8000/FormData`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(formData),
          }).then(() => {
            console.log(
              chalk.hex(`#29D479`)("Success, Form Data Added to Database")
            );
          });
        }
      }
    );
    formData.ReStatus = "Img";
    var webParts = [
      "Title",
      "Header",
      "Section1",
      "Section2",
      "Section3",
      // "Section4",
      "Section5",
      "Footer",
    ];
  }
  for (let webPart of webParts) {
    try {
      formData.reTry = 0;
      const statusPart = await webPartCreator(formData, webPart);
      wss.clients.forEach((client) => {
        if (client.readyState === WebSocket.OPEN) {
          client.send("reload");
        }
      });
      console.log(
        chalk.bold.magenta("Status for " + webPart + ": ") +
          chalk.green(statusPart)
      );
    } catch (error) {
      console.error(
        chalk.bold.magenta("Error occurred for " + webPart + " part: ") +
          chalk.red(error)
      );
    }
  }
  wss.clients.forEach((client) => {
    if (client.readyState === WebSocket.OPEN) {
      client.send("reload");
    }
  });
  console.log(chalk.hex("#932CAA").bold("Process Done"));
  res.status(200).json("OK");
});

// Calling API of AI(ChatGPT)
let requestQueue = [];
let lastRequestTime = 0;
app.post("/Call_Chat-GPT", async (req, res) => {
  try {
    const currentTime = Date.now();
    const elapsedSinceLastRequest = currentTime - lastRequestTime;
    if (elapsedSinceLastRequest < 10000 && requestQueue.length >= 3) {
      throw new Error(
        `Rate limit exceeded. Please wait before making another request.`
      );
    }
    requestQueue.push(currentTime);
    requestQueue = requestQueue.filter(
      (timestamp) => currentTime - timestamp <= 10000
    );
    lastRequestTime = currentTime;

    const data = req.body;
    const completion = await openai.createChatCompletion({
      model: "gpt-3.5-turbo",
      messages: [
        {
          role: "system",
          content:
            "Generate code for websites, put the code in this format: $HTML {code} !HTML , $CSS {code}[never include <style> tag here] !CSS , $JavaScript {code} !JavaScript, like they are on separate files and Use only English",
        },
        {
          role: "user",
          content: `${data.string}`,
        },
      ],
    });
    res.json({
      completion: completion.data.choices[0].message,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Calling API of AI(Dall-E)
app.post("/Call_Dall-E", async (req, res) => {
  try {
    const instructions = req.body;
    const response = await openai.createImage({
      prompt: `${instructions.string}`,
      n: 1,
      // size: "256x256",
      // size: "512x512",
      // size: "1024x1024",
      size: `${instructions.size}`,
      response_format: "b64_json",
    });
    res.json({
      response: response.data,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

server.listen(port, () => {
  console.log(`Server listening at http://localhost:${port}`);
});
